x/ccv/provider/keeper/grpc_query.go

```diff
@@ -270,31 +270,21 @@ func (k Keeper) QueryConsumerChainsValidatorHasToValidate(goCtx context.Context,
 
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
+	provAddr := types.NewProviderConsAddress(consAddr)
+
 	// get all the consumer chains for which the validator is either already
 	// opted-in, currently a consumer validator or if its voting power is within the TopN validators
 	consumersToValidate := []string{}
 	for _, consumer := range k.GetAllConsumerChains(ctx) {
 		chainID := consumer.ChainId
-		provAddr := types.NewProviderConsAddress(consAddr)
-		if !k.IsOptedIn(ctx, chainID, provAddr) && !k.IsConsumerValidator(ctx, chainID, provAddr) {
-			// check that the validator voting power isn't in the TopN
-			if topN, found := k.GetTopN(ctx, chainID); found && topN > 0 {
```

Comment:
The broken logic is that if this condition is false, then it should not continue further. Right?

Reply:
yes, but only part of the problem:
It also is problematic if the very first if on l279 is false, i.e. the val is not opted in and also not a consumer validator; then the if will never be entered, and it will just go to append the chainID.